mapper/thin_edge_json/src/json.rs

```diff
@@ -282,13 +291,94 @@ impl ThinEdgeJsonError {
     use super::*;
+    use crate::measurement::*;
 
     fn test_timestamp() -> DateTime<FixedOffset> {
         FixedOffset::east(5 * 3600)
             .ymd(2021, 04, 08)
             .and_hms(0, 0, 0)
     }
 
+    #[test]
+    fn valid_short_time_stamp() {
+        let input = r#"{
+            "time" : "2021-04-30T17:03:14.123+02:00"
+        }"#;
+        let json = json::parse(&input).unwrap();
+
+        let expected_output = FixedOffset::east(2 * 3600)
+            .ymd(2021, 04, 30)
+            .and_hms_milli(17, 03, 14, 123);
+
+        let time = json.entries().nth(0).unwrap();
+        let output = ThinEdgeJson::check_timestamp_for_iso8601_complaint(time.1).unwrap();
+        assert_eq!(output, expected_output);
+    }
+
+    #[test]
+    fn valid_string_time_stamp() {
+        let input = r#"{
+            "time" : "2021-04-30T17:03:14.123456789+02:00"
+        }"#;
+
+        let json = json::parse(&input).unwrap();
+        let expected_output = FixedOffset::east(2 * 3600)
+            .ymd(2021, 04, 30)
+            .and_hms_nano(17, 03, 14, 123456789);
+
+        let time = json.entries().nth(0).unwrap();
+        let output = ThinEdgeJson::check_timestamp_for_iso8601_complaint(time.1).unwrap();
+        assert_eq!(output, expected_output);
+    }
+
+    #[test]
+    fn invalid_time_stamp() {
+        let input = r#"{
+            "time" : "2013-06-2217:03:14.000658767+02:00"
+        }"#;
+        let expected_error = r#"Invalid ISO8601 timestamp (expected YYYY-MM-DDThh:mm:ss.sss.±hh:mm): "2013-06-2217:03:14.000658767+02:00": input contains invalid characters"#;
+        let json = json::parse(&input).unwrap();
+        let time = json.entries().nth(0).unwrap();
+        let output = ThinEdgeJson::check_timestamp_for_iso8601_complaint(time.1).unwrap_err();
+        assert_eq!(output.to_string(), expected_error);
+    }
+    #[test]
+    fn check_json_str_with_millisecond_timestamp() {
+        let input = r#"{
+            "time" : "2021-04-30T17:03:14.123+02:00",
+            "temperature" : 25
+        }"#;
+        let timestamp = FixedOffset::east(2 * 3600)
+            .ymd(2021, 04, 30)
+            .and_hms_milli(17, 03, 14, 123);
+
+        let output = ThinEdgeJson::from_str(input, timestamp).unwrap();
+        assert_eq!(
+            output.timestamp,
+            FixedOffset::east(2 * 3600)
+                .ymd(2021, 04, 30)
+                .and_hms_milli(17, 03, 14, 123)
+        );
+    }
+
+    #[test]
+    fn check_json_str_with_nanosecond_timestamp() {
+        let input = r#"{
+            "time" : "2021-04-30T17:03:14.123456789+02:00",
+            "temperature" : 25
+        }"#;
+
+        let timestamp = current_timestamp();
+
+        let output = ThinEdgeJson::from_str(input, timestamp).unwrap();
+        assert_eq!(
+            output.timestamp,
+            FixedOffset::east(2 * 3600)
+                .ymd(2021, 04, 30)
+                .and_hms_nano(17, 03, 14, 123456789)
+        );
+    }
+

```

Comment:
An additional test with no sub-seconds value would be good as all the existing unit tests are using decimal seconds values.

Reply:
added